llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/vec-args.ll

```diff
@@ -0,0 +1,1233 @@
+  ; RV64: bb.1.entry:
+  ; RV64-NEXT:   liveins: $x10
+  ; RV64-NEXT: {{  $}}
+  ; RV64-NEXT:   [[COPY:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY1:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY2:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY3:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY4:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY5:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY6:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY7:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY8:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY9:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY10:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY11:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY12:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY13:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY14:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY15:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY16:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY17:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY18:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY19:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY20:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY21:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY22:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY23:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY24:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY25:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY26:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY27:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY28:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY29:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY30:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[COPY31:%[0-9]+]]:_(s64) = COPY $x10
+  ; RV64-NEXT:   [[BUILD_VECTOR:%[0-9]+]]:_(<vscale x 32 x s64>) = G_BUILD_VECTOR [[COPY]](s64), [[COPY1]](s64), [[COPY2]](s64), [[COPY3]](s64), [[COPY4]](s64), [[COPY5]](s64), [[COPY6]](s64), [[COPY7]](s64), [[COPY8]](s64), [[COPY9]](s64), [[COPY10]](s64), [[COPY11]](s64), [[COPY12]](s64), [[COPY13]](s64), [[COPY14]](s64), [[COPY15]](s64), [[COPY16]](s64), [[COPY17]](s64), [[COPY18]](s64), [[COPY19]](s64), [[COPY20]](s64), [[COPY21]](s64), [[COPY22]](s64), [[COPY23]](s64), [[COPY24]](s64), [[COPY25]](s64), [[COPY26]](s64), [[COPY27]](s64), [[COPY28]](s64), [[COPY29]](s64), [[COPY30]](s64), [[COPY31]](s64)
+  ; RV64-NEXT:   [[TRUNC:%[0-9]+]]:_(<vscale x 32 x s16>) = G_TRUNC [[BUILD_VECTOR]](<vscale x 32 x s64>)
+  ; RV64-NEXT:   PseudoRET
+entry:
+  ret void
+}
+
+define void @test_args_nxv1b16(<vscale x 1 x bfloat> %a) {
+  ; RV32-LABEL: name: test_args_nxv1b16
+  ; RV32: bb.1.entry:
+  ; RV32-NEXT:   liveins: $f10_f
+  ; RV32-NEXT: {{  $}}
+  ; RV32-NEXT:   [[COPY:%[0-9]+]]:_(s32) = COPY $f10_f
+  ; RV32-NEXT:   [[BUILD_VECTOR:%[0-9]+]]:_(<vscale x 1 x s32>) = G_BUILD_VECTOR [[COPY]](s32)
```

Comment:
This looks wrong. A <vscale x 1 x bfloat> contains at least 1 scalar, but this G_BUILD_VECTOR assumes there is exactly 1 scalar.

Reply:
Our goal is to move a s32 which contains an extended s16 into a `<vscale x 1 x s16>`. I used G_BUILD_VECTOR to create a scalable vector with this element in it and since `<vscale x 1 x s16>` has enough room to hold at least one s16, then we can definitely hold the s16 were trying to put into it. 

If the vector ends up being bigger because vscale is greater than 1, I'm not sure that's a problem. How else do you recommend to fit the extended s32 into the `<vscale x 1 x s16>` the body of the program expects the argument to be lowered into?

Reply:
The caller would have also had a `<vscale x 1 x s32>` but only passed 1 value in f10_f to this function so the caller dropped the rest of the vector. That's bad.

In SelectionDAG, BUILD_VECTOR does not support scalable vectors.

In SelectinDAG, you pass a bfloat16 scalable vector without support for bf16 vectors the compiler crashes or throws an error. We could potentially bitcast it to a different scalable vector with the same total size, but we definitely cannot scalarlize it.

Reply:
> The caller would have also had a <vscale x 1 x s32> but only passed 1 value in f10_f to this function so the caller dropped the rest of the vector. That's bad.

IIUC, the problem is that the value is coming in as an s32 in the first place. In `determineArguments` we are getting back s32 from `TLI->getRegsiterTypeForCallingConv()`. We should fail at this point, or have failed before getting to this point with no f16 vector support similar to SDAG. WDYT?

How does this play with a previous comment:

> I'm not sure why we've singled out 16BitFPTy here. Aren't vectors of i128 or f128 also not supported or any other random integer type that's isn't i8/i16/i32/i64? Also vectors of i64 aren't supported with only the Zve32* extensions.

Do we need to error out if we're not one i8/i16/i32/i64/f64/f32? Do you think isSupportedArguments() the best place to do these checks?

Reply:
The caller would have also had a <vscale x 1 x s32> but only passed 1 value in f10_f to this function so the caller dropped the rest of the vector. That's bad.

IIUC, the problem is that the value is coming in as an s32 in the first place. In determineArguments we are getting back s32 from the RISCV::CC_RISCV function. We should fail at this point, or have failed before getting to this point with no f16 vector support similar to SDAG. WDYT?

How does this play with a previous comment:

I'm not sure why we've singled out 16BitFPTy here. Aren't vectors of i128 or f128 also not supported or any other random integer type that's isn't i8/i16/i32/i64? Also vectors of i64 aren't supported with only the Zve32* extensions.

Do we need to error out if we're not one i8/i16/i32/i64/f64/f32? Do you think isSupportedArguments() the best place to do these checks?

Reply:
> Do we need to error out if we're not one i8/i16/i32/i64/f64/f32? Do you think isSupportedArguments() the best place to do these checks?

I think isSupportedArgument() should replicate the checks from `RISCVTargetLowering::isLegalElementTypeForRVV` to ensure the element type is valid. bfloat16 seems to be missing there but that might be because isLegalElementTypeForRVV is also used in vectorizer cost modeling and I'm not sure if we're ready to vectorize bfloat16.